[WIP] Making 2FA viable for production.

mateuszmandera opened the pull request:
This isn't complete yet, but the remaining TODOs are reasonably simple or require additional discussion to figure out how we want things to work.

- [ ] Adding caching of `user_has_device` to make the last commit viable - otherwise it does database queries on user.is_authenticated calls. 
- [x] Fixing the monkey-patching of `TwoFactorLoginView` point of #9521 (we don't need to submit anything upstream as they've  merged some code that will allow us to do what we need here)
- [ ] Styling the HTML templates

Discussion is required on how this should work with mobile and zulip-terminal apps. What should be the right way of fetching a user's API key? Besides that, do we want to keep giving full API access with just the API key to users with 2FA enabled? Without some additional protections/restrictions, leaked API keys completely bypass the point of 2FA, so that might be considered a problem.

mateuszmandera committed:
auth: Remove redundant hack and comment from start_two_factor_auth.

Things appear work just fine without the hack. The comment is redundant,
as explaining usage of Django interfaces should be left to Django docs.

mateuszmandera committed:
auth: Add a basic 2fa auth test.

mateuszmandera committed:
auth: Don't set_password in 2fa with ldap test.

Given that this is testing ldap auth, setting a password for the
email+password backend (which is what user_profile.set_password) is
redundant and confusing.

mateuszmandera committed:
auth: Extract login_with_2fa test helper.

mateuszmandera committed:
auth: Use caching in OTPMiddleware.

The upstream implementation can't be used in production due to making
database query on every request by a user with 2FA enabled. Caching
needs to be used.

mateuszmandera committed:
auth: Disable the /fetch_api_key endpoint for users with 2fa.

Otherwise this endpoint is a wide-open backdoor completely bypassing the
protection of 2FA.

mateuszmandera committed:
auth: Integrate user_profile.is_authenticated with 2fa.

It's not reasonable to request.user.is_authenticated and
whether 2fa is enabled on the server + whether the user has 2FA
devices enable and if so, whether request.user.is_verified()
in every place where we want to ensure the user is authenticated.

If a user has 2fa enabled, we don't ever want to consider them
authenticated unless they're user.is_verified(). Thus, we customize the
is_authenticated property to be sufficient from our point of view.

timabbott commented:
> What should be the right way of fetching a user's API key? Besides that, do we want to keep giving full API access with just the API key to users with 2FA enabled? Without some additional protections/restrictions, leaked API keys completely bypass the point of 2FA, so that might be considered a problem.

Generally the way this works is (1) requiring 2FA to get an API key (i.e. mobile auth would need to support this), which we might implement through making mobile use the mobile_flow_otp process if necessary, and (2) having a list of configured API keys in the UI (#12511) where one can disable them.  It doesn't make sense to have the mobile app require a 2FA challenge every time one opens the app -- it'd be unuseable, break notifications, etc.


@gnprice FYI for the mobile question.

mateuszmandera commented:
>  It doesn't make sense to have the mobile app require a 2FA challenge every time one opens the app -- it'd be unuseable, break notifications, etc.

Agreed that it's not feasible. What came to my mind as perhaps a reasonable middle-ground is to have API keys with expiry. Aside of the normal API access, such a temporary API key could also allow renewing it by passing it + a valid 2FA code to an appropriate endpoint. With that, the mobile  app would only have to prompt the user for a 2FA key every X days.

mateuszmandera committed:
auth: Remove monkey-patching in TwoFactorLoginView.

andersk commented:
It’s usual practice for non-password tokens to stay valid more or less indefinitely as long as the client keeps using them. Although it’s also usual practice to

* give out multiple different tokens for a client with multiple devices and apps, so that tokens not in active use can be purged either automatically or manually;
* prohibit duplication of tokens from within the website;
* require reauthentication to generate new tokens.

With those safeguards, it should be fine for tokens to bypass 2FA.

I think we should consider replacing our token system with true OAuth, which was designed for this purpose. (With some transition period where we support both, of course.)

mateuszmandera committed:
auth: Rate limit 2fa token submission attempts.

6 digit OTP tokens are too short to be safe from brute-forcing and thus
some kind of limiting of attempts is necessary.

https://lukeplant.me.uk/blog/posts/6-digit-otp-for-two-factor-auth-is-brute-forceable-in-3-days/
has a nice elaboration on this problem. It also specifically talks about
django-two-factor-auth and django-otp which we rely on. Those used to
not have the necessary protections, but a throttling mechanism was
implemented in
https://github.com/django-otp/django-otp/commit/fab7199d47a7d17a75f288fb5161dc44b8c7fe31
but django-two-factor-auth doesn't properly take advantage of it. We
cannot directly use the mechanism ourselves, as it involves applying
ThrottlingMixin to appropriate Device classes - those are defined in
django-two-factor-auth code however and thus we don't have a good way of
customizing them in this way. We can achieve the desired outcome through
a straight-forward application of our rate limiting system however.

https://github.com/Bouke/django-two-factor-auth/issues/299 is a relevant
upstream issue that we can track to know when we might be able to clean
up this code. On the other hand, we might want to keep doing things our
own way via out rate-limiting system as it leaves less room for some
gaps in upstream code (such as a Device class that should be throttled,
but isn't, going unnoticed).
This can be reconsidered in the future.

mateuszmandera committed:
auth: Make 2fa html templates overridable.

The small change to TwoFactorLoader allows overriding of two_factor
templates by defining our own in templates/two_factor/.

We do this on the _base.html template, which we end up just copy-pasting
with an alert removed, to have some very barebones, but usable pages.
They will need to be styled, but there is a significant amount of
complication involved in doing it properly - the underlying library
writes all its default templates in Django's template language rather
than Jinja2. This meams that it's hard to re-use our webpack-processed
CSS files etc. to have consistent styling.

mateuszmandera commented:
`django-two-factor-auth` using Django template language is a problem right now  - I see 3 options:
1. Try to figure out some hacky way of making the DTL templates use our webpacked CSS (I don't really have an ideas here though right now)
2. Work with the DTL templates and create some styling for it despite not being able to webpack the CSS or use our normal CSS.
3. Rewrite all of their templates with our own, Jinja2-based templates

gnprice commented:
> It’s usual practice for non-password tokens to stay valid more or less indefinitely as long as the client keeps using them. Although it’s also usual practice to
> 
> * give out multiple different tokens for a client with multiple devices and apps, so that tokens not in active use can be purged either automatically or manually;
> * prohibit duplication of tokens from within the website;
> * require reauthentication to generate new tokens.
> 
> With those safeguards, it should be fine for tokens to bypass 2FA.

I agree with all this.

Those three together make a security feature that we've wanted for a few years. #12511 was some partial work in that direction. It looks like we don't have a GitHub issue for this desired feature; but here's where we've previously discussed the design for it in chat, in 2018 and then 2019:
https://chat.zulip.org/#narrow/stream/3-backend/topic/new.20api.20key.20after.20pass.20change/near/609933
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/multiple.20API.20keys/near/787516 

That feature would be helpful for Zulip users' security even independent of 2FA, but 2FA gives a strong new reason to complete it.

gnprice commented:
Oh, and I meant to add:

> I think we should consider replacing our token system with true OAuth, which was designed for this purpose. (With some transition period where we support both, of course.)

This sounds plausible to me but I don't really know anything about the details of what it would mean.

gnprice commented:
I've now filed #17939 for the prerequisite feature Anders and I described above.

timabbott commented:
The first 3-4 commits are clean refactors that certainly don't need to block on the above; I'd be happy to merge those in the next couple days after a rebase (as they appear to all merge conflict :( ).
